### PR TITLE
new zfs live ebuilds

### PR DESCRIPTION
**Pull request by gyakovlev**
I've done a some work on live ebuilds.

I'd like to have some feedback and testing from dev using zfs.
I tried contacting @ryao several times using irc/email but he's quite busy and unresponsive.

live and 0.8.0 ebuilds also need to install pyzfs, I haven't decided if it's going to be a useflag or separate package.

I also would like to get rid of our systemd unit in `FILESDIR` and just use upstream one, for 9999 and 0.8 ebuilds, many changes went into  systemd integration upstream.

since zfs/spl 9999 ebuilds tend to get outdated  I based my work on 0.7.12, not on 9999 ebuilds.

@prometheanfire @rich0 @leio @perfinion tagging you guys as I know you are zfs users

if proven to work I'd like to have 0.7.13 to be based on those ebuilds, but it can wait until 0.8.0 is released.

### CODE REVIEW

File: sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild [hunk 1]
```diff
@@ -63,64 +66,90 @@ pkg_setup() {
 			DEVTMPFS
 	"
 
-	kernel_is ge 2 6 32 || die "Linux 2.6.32 or newer required"
+	kernel_is -lt 5 && CONFIG_CHECK="${CONFIG_CHECK} IOSCHED_NOOP"
+
+	kernel_is -ge 2 6 32 || die "Linux 2.6.32 or newer required"
+
+	local KV_MAJOR_MAX="${ZFS_KERNEL_COMPAT%%.*}"
```
[the remark this answers is not in the document]
Comment (gyakovlev) [reply sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild]: at the time it was 4.19 compatible, will update

File: sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild [hunk 2]
```diff
@@ -63,64 +66,90 @@ pkg_setup() {
...
+src_compile() {
+	set_arch_to_kernel
+	default
 }
 
 src_install() {
-	autotools-utils_src_install INSTALL_MOD_PATH="${INSTALL_MOD_PATH:-$EROOT}"
+	set_arch_to_kernel
+	emake \
+		DEPMOD="/bin/true" \
```
[the remark this answers is not in the document]
Comment (gyakovlev) [reply sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild]: because `linux-mod.eclass` handles that.

[the remark this answers is not in the document]
Comment (gyakovlev) [reply sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild]: namely 
```
linux-mod_pkg_postinst
              It executes /sbin/depmod and adds the package to the /var/lib/module-rebuild/moduledb database (if ${D}/lib/modules is created)"

```

File: sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild [hunk 3]
```diff
@@ -1,52 +1,55 @@
...
 else
 	SRC_URI="https://github.com/zfsonlinux/zfs/releases/download/zfs-${PV}/zfs-${PV}.tar.gz"
+	KEYWORDS="~amd64"
 	S="${WORKDIR}/zfs-${PV}"
-	KEYWORDS="~amd64 ~arm ~ppc ~ppc64"
+	ZFS_KERNEL_COMPAT="${ZFS_KERNEL_COMPAT_OVERRIDE:-4.19}"
 fi
 
-inherit flag-o-matic linux-info linux-mod toolchain-funcs autotools-utils
+if [[ ${PV} == "0.7"* ]]; then
```
[the remark this answers is not in the document]
Comment (gyakovlev) [reply sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild]: this is do differentiate with 9999 and 0.7.x, including 0.7.9999, not 0.6.x

**Comment by gyakovlev** [comment]
> looks like the main diff between 0.7.9999 and 9999 for zfs-kmod are the kernel options?

and absence of spl dep and different docs.
the intention was to keep ebuilds in sync and not let them rot.
> 
> I think this should be approved by one of the actual maintainers of the ebiulds though. This is a VERY large changeset to do all at once.

I would like to get some feedback from @ryao but he's been mia for some time.
not pushing this very hard, but we'll have to do that at some point.

File: sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild (hunk 2 above)
[the remark this answers is not in the document]
Comment (prometheanfire) [reply sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild]: k

File: sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild (hunk 3 above)
[the remark this answers is not in the document]
Comment (prometheanfire) [reply sys-fs/zfs-kmod/zfs-kmod-0.7.9999.ebuild]: ok

**Comment by prometheanfire** [comment]
ok, that's fine.

I'd be more willing to merge a smaller changeset myself, can you limit it to 9999 for now?  once that sits for a little bit, we can update 0.7.9999 and the like.

**Comment by gyakovlev** [comment]
sure, I'll remove all 0.7.x specific conditional code and ebuilds from here and just target 9999 ebuilds which should be used as a base futher 0.8.x
0.7.9999 should be  separate live ebuilds for simplicity, too many changes happened to handle branch differences within single ebuild.

**Comment by prometheanfire** [comment]
ya, I'd consider one template per branch to make sense.

**Comment by gyakovlev** [comment]
ok I've redone everything, just 9999 ebuild for now.
we can have pyzfs, for all python implementations.
now using upstream systemd units, as there was a conflict between our zfs.service and zfs-mount.service from upstream.
and many more small changes.

**Comment by prometheanfire** [comment]
added ryao to this as another reviewer

**Comment by gyakovlev** [comment]
trim support just landed, I guess 0.8.x release is getting ready soon and it'll be so nice to base 0.8 ebuild on modern EAPI.

I've been using this set of ebuilds with various set of repos  and branches (thanks to git-r3 OVERRIDE stuff) with no problems at all.

**Comment by prometheanfire** [comment]
testing this now, if it works I'll merge it, @ryao has had enough time to say no

**Comment by gyakovlev** [comment]
I can merge myself ;-) just wanted some feedback and wider testing before I do, in case I missed something.
one thing I didn't test is `kernel-builtin` though.

**Comment by prometheanfire** [comment]
well, I do that, so we're good there :D

**Comment by gyakovlev** [comment]
good! lemme know how it goes.
I've been postponing own migration to 9999 for a while but there are so many nice features in master compared to 0.7, can't go back. now I need to try built-in.

**Comment by prometheanfire** [comment]
it worked, feel free to merge

**Comment by gyakovlev** [comment]
thanks for testing!
merged

**Pull request closed without merging**